src/doc/algorithm/shift_image.cpp

```diff
@@ -20,31 +23,101 @@
+void shift_image(Image* image, int dx, int dy, double angle)
 {
-  gfx::Rect bounds = mask->bounds();
-  if (bounds.isEmpty())
-    return;
-
+  gfx::Rect bounds(image->bounds());
+  if (cos(angle) < -sqrt(2)/2) {
+    dx = -dx;
+    dy = -dy;
+  }
+  else if (sin(angle) >= sqrt(2)/2) {
+    double aux;
+    aux = dx;
+    dx = -dy;
+    dy = aux;
+  }
+  else if (sin(angle) < -sqrt(2)/2) {
+    double aux;
+    aux = dx;
+    dx = dy;
+    dy = -aux;
+  }
   // To simplify the algorithm we use a copy of the original image, we
   // could avoid this copy swapping rows and columns.
   ImageRef crop(crop_image(image, bounds.x, bounds.y, bounds.w, bounds.h,
-                           image->maskColor()));
-
-  int u = dx;
-  int v = dy;
-  while (u < 0) u += bounds.w;
-  while (v < 0) v += bounds.h;
+                         image->maskColor()));
 
   for (int y=0; y<bounds.h; ++y) {
     for (int x=0; x<bounds.w; ++x) {
-      put_pixel(image,
-                bounds.x + ((u + x) % bounds.w),
-                bounds.y + ((v + y) % bounds.h),
-                get_pixel(crop.get(), x, y));
+        put_pixel(image,
+                  (bounds.w + dx + x) % bounds.w,
+                  (bounds.h + dy + y) % bounds.h,
+                  get_pixel(crop.get(), x, y));
     }
   }
 }
 
+void shift_image_with_mask(Cel* cel, const Mask* mask, int dx, int dy)
+{
+  ASSERT(!cel->bounds().isEmpty());
+  ASSERT(!mask->bounds().isEmpty());
+
+  // Making a image which bounds are equal to the UNION of cel->bounds() and mask->bounds()
+  gfx::Rect compCelBounds = cel->bounds() | mask->bounds();
+  ImageRef compImage(Image::create(cel->image()->pixelFormat(), compCelBounds.w, compCelBounds.h));
+
+  // Making a Rect which represents the mask bounds of the original MASK relative to
+  // the new COMPOUND IMAGE (compImage)
+  gfx::Point maskCelGap(0, 0);
+  if (cel->bounds().x < mask->bounds().x)
+    maskCelGap.x = mask->bounds().x - cel->bounds().x;
+  else
+    maskCelGap.x = 0;
+  if (cel->bounds().y < mask->bounds().y)
+    maskCelGap.y = mask->bounds().y - cel->bounds().y;
+  else
+    maskCelGap.y = 0;
+  gfx::Rect maskedBounds(maskCelGap.x, maskCelGap.y, mask->bounds().w, mask->bounds().h);
+
+  // Making one combined image: Image with the Mask Bounds (unfilled spaces were filled with mask color)
+  compImage->copy(cel->image(), gfx::Clip(cel->position().x-compCelBounds.x,
+                                          cel->position().y-compCelBounds.y,
+                                          0, 0,
+                                          cel->bounds().w, cel->bounds().h));
+
+  // Making a copy of only the image which will be shiftted
+  ImageRef imageToShift(Image::create(compImage->pixelFormat(), maskedBounds.w, maskedBounds.h));
+  imageToShift->copy(compImage.get(), gfx::Clip(0, 0, maskedBounds.x, maskedBounds.y,
+                                                imageToShift->bounds().w, imageToShift->bounds().h));
```

Comment:
This can be simplified:
```c++
imageToShift->copy(compImage.get(), gfx::Clip(0, 0, maskedBounds));
```
Because `imageToShift->bounds().w == maskedBounds.w` and `imageToShift->bounds().h == maskedBounds.h`